Add Web Extension code generator scripts, bindings glue, and some IDL files.

xeenon opened the pull request:
#### 0d554516d6f8d662d907d5200b95b9c321e29d62
<pre>
Add Web Extension code generator scripts, bindings glue, and some IDL files.
<a href="https://bugs.webkit.org/show_bug.cgi?id=246410">https://bugs.webkit.org/show_bug.cgi?id=246410</a>

Reviewed by Brian Weinstein.

This stands up the code generator for the JavaScript bindings for Web Extensions.
This has been tested with all the IDL files, but I&apos;m only adding a couple partial
interfaces with stub implementations to keep this change small. The rest of the
interfaces files will be added as the API areas are added.

For each IDL file in WebProcess/Extensions/Interfaces, a corresponding header and
implementation is expected in WebProcess/Extensions/API.

The bulk of the binding helpers are defined in JSWebExtensionWrapper.h, with converter
functions and the JSWebExtensionWrapper and WebExtensionCallbackHandler classes.

The WebExtensionCallbackHandler class handles the two ways Web Extensions can
process async results — the Manifest v2 callback function way, or the Manifest v3
Promise result method. This lets the API layer be agnostic about which one the
JavaScript is using and report results or errors consistently.

This is untested (besides successfully building) until API injection can be done.
Most of this was copied from Safari&apos;s existing implementation and changed for
WebKit to use more C++, and use WebCore&apos;s bindings scripts.

* Source/WebKit/DerivedSources-input.xcfilelist: Auto generated.
* Source/WebKit/DerivedSources-output.xcfilelist: Ditto.
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/generate-derived-sources.sh: Added WebCoreScripts var.
* Source/WebKit/SourcesCocoa.txt: Added new file.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Added new files.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIExtensionCocoa.mm: Added.
(WebKit::WebExtensionAPIExtension::isPropertyAllowed):
(WebKit::WebExtensionAPIExtension::getURL):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm: Added.
(WebKit::WebExtensionAPINamespace::extension):
(WebKit::WebExtensionAPINamespace::runtime):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm: Added.
(WebKit::WebExtensionAPIRuntimeBase::reportErrorForCallbackHandler):
(WebKit::WebExtensionAPIRuntime::getURL):
(WebKit::WebExtensionAPIRuntime::getManifest):
(WebKit::WebExtensionAPIRuntime::runtimeIdentifier):
(WebKit::WebExtensionAPIRuntime::getPlatformInfo):
(WebKit::WebExtensionAPIRuntime::lastError):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIExtension.h: Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h: Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIObject.h: Added.
(WebKit::WebExtensionAPIObject::WebExtensionAPIObject):
(WebKit::WebExtensionAPIObject::forMainWorld const):
(WebKit::WebExtensionAPIObject::isForMainWorld const):
(WebKit::WebExtensionAPIObject::runtime):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIRuntime.h: Added.
* Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm: Added.
(WebKit::WebExtensionCallbackHandler::WebExtensionCallbackHandler):
(WebKit::WebExtensionCallbackHandler::~WebExtensionCallbackHandler):
(WebKit::WebExtensionCallbackHandler::callbackFunction const):
(WebKit::callWithArguments):
(WebKit::WebExtensionCallbackHandler::reportError):
(WebKit::WebExtensionCallbackHandler::call):
(WebKit::functionObjectByName):
(WebKit::toNSObject):
(WebKit::toNSString):
(WebKit::toNSDictionary):
(WebKit::toJSValue):
(WebKit::toJSCallbackHandler):
(WebKit::deserializeJSONString):
(WebKit::serializeJSObject):
(-[JSValue _awaitThenableResolutionWithCompletionHandler:]):
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrappable.h: Added.
(WebKit::JSWebExtensionWrappable::~JSWebExtensionWrappable):
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.cpp: Added.
(WebKit::JSWebExtensionWrapper::wrap):
(WebKit::JSWebExtensionWrapper::unwrap):
(WebKit::unwrapObject):
(WebKit::JSWebExtensionWrapper::initialize):
(WebKit::JSWebExtensionWrapper::finalize):
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.h: Added.
(WebKit::WebExtensionCallbackHandler::create):
(WebKit::WebExtensionCallbackHandler::globalContext const):
(WebKit::toWebFrame):
(WebKit::toWebPage):
(WebKit::toJSString):
(WebKit::toJSNullIfNull):
(WebKit::toJS):
(WebKit::toJSPromiseCallbackHandler):
(WebKit::toJSErrorCallbackHandler):
(WebKit::toJSValue):
(WebKit::toJSError):
* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/CodeGeneratorExtensions.pm: Added.
* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/IDLAttributes.json: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIExtension.idl: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIRuntime.idl: Added.
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp: Build fix for Sources.txt change.
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp: Ditto.
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp: Ditto.

Canonical link: <a href="https://commits.webkit.org/256120@main">https://commits.webkit.org/256120@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c691505a719da410475414e1249689ce150dfc3e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94755 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3911 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/27632 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104369 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3982 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/32094 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/87064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/100293 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100422 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/2866 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/81112 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/87064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/98341 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/84794 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/84340 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/87064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38503 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36342 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/19434 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/4224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40260 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42080 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2018 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42234 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38665 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/DerivedSources.make:
```diff
@@ -326,6 +327,14 @@ GENERATE_MESSAGE_RECEIVER_SCRIPTS = \
+
+PLATFORM_HEADER_DIR := $(realpath $(BUILT_PRODUCTS_DIR)$(WK_LIBRARY_HEADERS_FOLDER_PATH))
+PLATFORM_HEADER_DEPENDENCIES := $(filter $(PLATFORM_HEADER_DIR)/%,$(realpath $(shell $(call platform_h_compiler_command,-M) | $(PERL) -e "local \$$/; my (\$$target, \$$deps) = split(/:/, <>); print split(/\\\\/, \$$deps);")))
+FEATURE_AND_PLATFORM_DEFINE_DEPENDENCIES = $(WebKit2)/DerivedSources.make $(PLATFORM_HEADER_DEPENDENCIES)
```
I'm going to take your word for this one :).

xeenon replied:
Hah! That came from WebCore's version, so I take no credit.

b-weinstein commented on Source/WebKit/DerivedSources.make:
```diff
@@ -552,3 +561,27 @@ all : GeneratedSerializers.h GeneratedSerializers.cpp GeneratedSerializers.mm Se
+    $(WebCoreScripts)/generate-bindings.pl \
+    $(WebCoreScripts)/IDLParser.pm \
+    $(WebCoreScripts)/CodeGenerator.pm \
+    $(EXTENSIONS_SCRIPTS_DIR)/CodeGeneratorExtensions.pm \
```
I think this would be slightly more readable as ExtensionCodeGenerator, but if other code generators use this pattern (or even if it's a pain to fix), I don't feel strongly.

xeenon replied:
Yeah this is the naming pattern for all the other generators. I don't think it technically matters, but I kept it.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIExtensionCocoa.mm:
```diff
@@ -0,0 +1,51 @@
+bool WebExtensionAPIExtension::isPropertyAllowed(String, WebPage*)
+{
+    // FIXME: Implement.
+    return true;
```
Super-nit: I'd say we could `return false` for safety until this is implemented.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
```diff
@@ -0,0 +1,53 @@
+WebExtensionAPIRuntime& WebExtensionAPINamespace::runtime()
+{
+    if (!m_runtime)
+        m_runtime = WebExtensionAPIRuntime::create(forMainWorld());
```
Will we need to support these APIs being injected in content scripts here too? Or will that be a separate class?

b-weinstein replied:
Oh I see, forMainWorld() is a bool method.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -0,0 +1,112 @@
+    callback.call();
+
+    if (!m_lastErrorAccessed) {
+        // Long the error to the console if it wasn't checked in the callback.
```
Typo: Log.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm:
```diff
@@ -0,0 +1,351 @@
+        toJSValue(m_globalContext.get(), argumentOne),
+        toJSValue(m_globalContext.get(), argumentTwo),
+        toJSValue(m_globalContext.get(), argumentThree)
+    });
```
I'm a little surprised there aren't helper methods like this anywhere in the tree.

xeenon replied:
It is the wild C++ west sometimes. But this is almost verbatim from the Safari copy.

xeenon committed:
Add Web Extension code generator scripts, bindings glue, and some IDL files.
https://bugs.webkit.org/show_bug.cgi?id=246410

Reviewed by Brian Weinstein.

This stands up the code generator for the JavaScript bindings for Web Extensions.
This has been tested with all the IDL files, but I'm only adding a couple partial
interfaces with stub implementations to keep this change small. The rest of the
interfaces files will be added as the API areas are added.

For each IDL file in WebProcess/Extensions/Interfaces, a corresponding header and
implementation is expected in WebProcess/Extensions/API.

The bulk of the binding helpers are defined in JSWebExtensionWrapper.h, with converter
functions and the JSWebExtensionWrapper and WebExtensionCallbackHandler classes.

The WebExtensionCallbackHandler class handles the two ways Web Extensions can
process async results — the Manifest v2 callback function way, or the Manifest v3
Promise result method. This lets the API layer be agnostic about which one the
JavaScript is using and report results or errors consistently.

This is untested (besides successfully building) until API injection can be done.
Most of this was copied from Safari's existing implementation and changed for
WebKit to use more C++, and use WebCore's bindings scripts.

* Source/WebKit/DerivedSources-input.xcfilelist: Auto generated.
* Source/WebKit/DerivedSources-output.xcfilelist: Ditto.
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/generate-derived-sources.sh: Added WebCoreScripts var.
* Source/WebKit/SourcesCocoa.txt: Added new file.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Added new files.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIExtensionCocoa.mm: Added.
(WebKit::WebExtensionAPIExtension::isPropertyAllowed):
(WebKit::WebExtensionAPIExtension::getURL):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm: Added.
(WebKit::WebExtensionAPINamespace::extension):
(WebKit::WebExtensionAPINamespace::runtime):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm: Added.
(WebKit::WebExtensionAPIRuntimeBase::reportErrorForCallbackHandler):
(WebKit::WebExtensionAPIRuntime::getURL):
(WebKit::WebExtensionAPIRuntime::getManifest):
(WebKit::WebExtensionAPIRuntime::runtimeIdentifier):
(WebKit::WebExtensionAPIRuntime::getPlatformInfo):
(WebKit::WebExtensionAPIRuntime::lastError):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIExtension.h: Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h: Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIObject.h: Added.
(WebKit::WebExtensionAPIObject::WebExtensionAPIObject):
(WebKit::WebExtensionAPIObject::forMainWorld const):
(WebKit::WebExtensionAPIObject::isForMainWorld const):
(WebKit::WebExtensionAPIObject::runtime):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIRuntime.h: Added.
* Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm: Added.
(WebKit::WebExtensionCallbackHandler::WebExtensionCallbackHandler):
(WebKit::WebExtensionCallbackHandler::~WebExtensionCallbackHandler):
(WebKit::WebExtensionCallbackHandler::callbackFunction const):
(WebKit::callWithArguments):
(WebKit::WebExtensionCallbackHandler::reportError):
(WebKit::WebExtensionCallbackHandler::call):
(WebKit::functionObjectByName):
(WebKit::toNSObject):
(WebKit::toNSString):
(WebKit::toNSDictionary):
(WebKit::toJSValue):
(WebKit::toJSCallbackHandler):
(WebKit::deserializeJSONString):
(WebKit::serializeJSObject):
(-[JSValue _awaitThenableResolutionWithCompletionHandler:]):
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrappable.h: Added.
(WebKit::JSWebExtensionWrappable::~JSWebExtensionWrappable):
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.cpp: Added.
(WebKit::JSWebExtensionWrapper::wrap):
(WebKit::JSWebExtensionWrapper::unwrap):
(WebKit::unwrapObject):
(WebKit::JSWebExtensionWrapper::initialize):
(WebKit::JSWebExtensionWrapper::finalize):
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.h: Added.
(WebKit::WebExtensionCallbackHandler::create):
(WebKit::WebExtensionCallbackHandler::globalContext const):
(WebKit::toWebFrame):
(WebKit::toWebPage):
(WebKit::toJSString):
(WebKit::toJSNullIfNull):
(WebKit::toJS):
(WebKit::toJSPromiseCallbackHandler):
(WebKit::toJSErrorCallbackHandler):
(WebKit::toJSValue):
(WebKit::toJSError):
* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/CodeGeneratorExtensions.pm: Added.
* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/IDLAttributes.json: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIExtension.idl: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIRuntime.idl: Added.
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp: Build fix for Sources.txt change.
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp: Ditto.
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp: Ditto.

Canonical link: https://commits.webkit.org/256120@main
